ddspipe_core/src/cpp/monitoring/producers/TopicsMonitorProducer.cpp

```diff
@@ -0,0 +1,333 @@
+    period = configuration.period;
+
+    // TODO: initialize the consumers outside of the producer so it remains agnostic of its consumers.
+
+    // Register the type
+    fastdds::dds::TypeSupport type(new MonitoringTopicsPubSubType());
+
+    // Register the type object
+    registerMonitoringTopicsTypes();
+
+    // Create the consumers
+    consumers_.push_back(std::make_unique<DdsMonitorConsumer<MonitoringTopics>>(
+                configuration.domain.get_value(), configuration.topic_name, type));
+    consumers_.push_back(std::make_unique<StdoutMonitorConsumer<MonitoringTopics>>());
+}
+
+void TopicsMonitorProducer::consume()
+{
+    if (!enabled_)
+    {
+        // Don't consume if the producer is not enabled
+        return;
+    }
+
+    const auto data = save_data_();
+
+    for (auto& consumer : consumers_)
+    {
+        consumer->consume(&data);
+    }
+
+    reset_data_();
+}
+
+void TopicsMonitorProducer::msg_received(
+        const types::DdsTopic& topic,
+        const types::ParticipantId& participant_id)
+{
+    if (!enabled_)
+    {
+        // Don't save the data if the producer is not enabled
+        return;
+    }
+
+    // Take the lock to prevent:
+    //      1. Changing the data while it's being saved.
+    //      2. Simultaneous calls to msg_received.
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    const auto& topic_name = topic.m_topic_name;
+    const auto& type_name = topic.type_name;
+
+    if (!topic_data_.count(topic_name))
+    {
+        topic_data_[topic_name].name(topic_name);
+        topic_data_[topic_name].type_name(type_name);
+    }
+
+    if (!participant_data_.count(topic_name) || !participant_data_[topic_name].count(participant_id))
+    {
+        // First message received for topic & participant
+        // Save the participant_id
+        participant_data_[topic_name][participant_id].participant_id(participant_id);
+    }
+
+    // Increase the count of the received messages
+    participant_data_[topic_name][participant_id].msgs_received(participant_data_[topic_name][participant_id].msgs_received() +
+            1);
+}
+
+void TopicsMonitorProducer::msg_lost(
+        const types::DdsTopic& topic,
+        const types::ParticipantId& participant_id)
+{
+    if (!enabled_)
+    {
+        // Don't save the data if the producer is not enabled
+        return;
+    }
+
+    // Take the lock to prevent:
+    //      1. Changing the data while it's being saved.
+    //      2. Simultaneous calls to msg_lost.
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    const auto& topic_name = topic.m_topic_name;
+    const auto& type_name = topic.type_name;
+
+    if (!participant_data_.count(topic_name) || !participant_data_[topic_name].count(participant_id))
+    {
+        // First message lost for topic & participant
+        // Save the participant_id
+        participant_data_[topic_name][participant_id].participant_id(participant_id);
+    }
```

Comment:
I think you can remove these checks everywhere, less code and I believe a tiny bit more efficient.